SSA/Core/Framework.lean

```diff
@@ -618,68 +1499,289 @@ theorem HVector.map_eq_of_eq_on_vars {A : Ty → Type*}
-    . rcases e with ⟨op, rfl, args⟩
-      simp [denote, Expr.denote]
-      congr 1
-      apply HVector.map_eq_of_eq_on_vars
-      intro v h'
-      apply ih
-      intro v' hv'
-      apply h
-      rw [vars, Var.casesOn_last]
-      simp
-      use v.1, v.2
+-- theorem Lets.denote_eq_of_eq_on_vars_aux [LawfulMonad m] (lets : Lets Op Γ_in eff Γ_out)
+--     (v : Var Γ_out t)
+--     {s₁ s₂ : Valuation Γ_in}
+--     (h : ∀ w, w ∈ lets.vars v → s₁ w.2 = s₂ w.2)
+--     {α} (k : Valuation Γ_out → eff.toMonad m α) :
+--     --   ^ represents arbitrary side effecting kontinuation that may happen in between
+--     (do
+--       let Γv ← lets.denote s₁
+--       let _x ← k Γv
+--       return Γv v)
+--     = (do
+--       let Γv ← lets.denote s₂
+--       let _x ← k Γv
+--       return Γv v) := by
+--   induction lets generalizing t
+--   case nil =>
+--     simp [vars] at h
+--     simp [denote, h, map_pure, Var.denote]
+--   case lete Γ_in eff Γ_out lets e body ih =>
+--     cases v using Var.casesOn
+--     . simp [vars] at h
+--       simp [denote]
+--       cases eff
+--       case pure =>
+--         simp_all
+--         apply ih
+--         simpa
+--       case impure =>
+--         simp_all [map_pure]
+--         sorry
+--         -- This needs some thought about side effects. Pretty sure it's true still, but it needs thinking.
+--         /-
+--         apply ih
+--         simpa
+--         -/
+--     . rcases e with ⟨op, rfl, args⟩
+--       sorry
+--       sorry
+--       /-
+--       simp [denote, Expr.denote]
+--       congr 1
+--       apply HVector.map_eq_of_eq_on_vars
+--       intro v h'
+--       apply ih
+--       intro v' hv'
+--       apply h
+--       rw [vars, Var.casesOn_last]
+--       simp
+--       use v.1, v.2
+--       -/
+
+-- /--
+--   Fix a variable v.
+--   If the two value1, s2, agree on all of values v uses in v's def-use chain,
+--   then the value of v after denoting the program is also equal.
+
+--   This means that we never need to look at the whole context, but only the fragment
+--   of the context that contains the expression tree of v.
+--   TODO: unused theorem?
+--   TODO: this theorem is not true in the impure case: the side-effects that occur outside of the
+--     def-use chain for variable `v` might depend on some variable of the valuations that is not part
+--     of `lets.vars v`, and might thus differ between these valuations.
+--     We could fix it by saying the valuations need to agree on all variables that are used in `lets`,
+--     but maybe it's not worth it, given that the theorem is unused
+--   -/
+-- theorem Lets.denote_eq_of_eq_on_vars [LawfulMonad m] (lets : Lets Op Γ_in eff Γ_out)
+--     (v : Var Γ_out t)
+--     {s₁ s₂ : Valuation Γ_in}
+--     (h : ∀ w, w ∈ lets.vars v → s₁ w.2 = s₂ w.2) :
+--     (lets.denote s₁) >>= (fun Γv => return v.denote Γv) =
+--     (lets.denote s₂) >>= (fun Γv => return v.denote Γv) := by
+--   simpa using denote_eq_of_eq_on_vars_aux lets v h pure
```

Comment:
Why are we keeping this around? (It might be worth having the counterexample be an `example` and made explicit, I guess?)